src/brod_transaction.erl

```diff
@@ -0,0 +1,343 @@
+produce_cb(Transaction, Topic, Partition, Key, Value, AckCb) ->
+  {ok, ProducerPid} = gen_server:call(Transaction, {get_producer, Topic, Partition}),
+  brod_producer:produce_cb(ProducerPid, Key, Value, AckCb).
+
+%% @see brod_producer:sync_produce_request/2
+-spec sync_produce_request(call_ref(), timeout()) ->
+  {ok, offset()} | {error, Reason}
+    when Reason :: timeout | {producer_down, any()}.
+sync_produce_request(CallRef, Timeout) ->
+  brod_producer:sync_produce_request(CallRef, Timeout).
+
+%% @doc adds the offset consumed by a group to the transaction.
+-spec add_offsets(transaction(), group_id(), offsets_to_commit()) -> ok | {error, any()}.
+add_offsets(Transaction, ConsumerGroup, Offsets) ->
+  gen_server:call(Transaction, {add_offsets, ConsumerGroup, Offsets}).
+
+%% @doc commits the transaction, after this, the gen_server will stop
+-spec commit(transaction()) -> ok | {error, any()}.
+commit(Transaction) ->
+  gen_server:call(Transaction, commit).
+
+%% @doc aborts the transaction, after this, the gen_server will stop
+-spec abort(transaction()) -> ok | {error, any()}.
+abort(Transaction) ->
+  gen_server:call(Transaction, abort).
+
+%% @doc aborts the transaction with a timeout, after this, the gen_server will
+%% stop
+-spec abort(transaction(), pos_integer()) -> ok | {error, any()}.
+abort(Transaction, Timeout) ->
+  gen_server:call(Transaction, {abort, Timeout}).
+
+%% @doc stops the transaction.
+-spec stop(transaction()) -> ok | {error, any()}.
+stop(Transaction) ->
+  gen_server:call(Transaction, terminate).
+
+%% @private
+make_transactional_id() ->
+  iolist_to_binary([atom_to_list(?MODULE), "-txn-",
+                    base64:encode(crypto:strong_rand_bytes(8))]).
+
+%% @private
+make_txn_context(Client, TxId)->
+  {ok, {{Host, Port}, _}} = brod_client:get_transactional_coordinator(Client,
+                                                                      TxId),
+  {ok, Conn} = brod_client:get_connection(Client, Host, Port),
+
+  kpro:txn_init_ctx(Conn, TxId).
+
+%% @private
+-spec may_add_producer(topic(), partition(), state()) -> state().
+may_add_producer(Topic, Partition,
+                 #state{ client_pid = ClientPid
+                       , producers = OldProducers
+                       , context = CTX
+                       , producers_config = ProducersConfig
+                       } = OldState) ->
+  case maps:is_key({Topic, Partition}, OldProducers) of
+    true -> OldState;
+    _ ->
+      {ok, ProducerPid} = brod_producer:start_link(ClientPid,
+                                                   Topic,
+                                                   Partition,
+                                                   CTX,
+                                                   ProducersConfig),
+
+      %% before producing something in a topic partition, we have to
+      %% announce it
+      ok = kpro:txn_send_partitions(CTX, [{Topic, Partition}]),
```

Comment:
I'm not sure if this can be done for the same topic-partition more than once in a transaction.
This is how I understand the purpose of this: For each transaction, Kafka needs to record the start point offsets for each topic-partition.

If the 'begin' mark for each topic-partition can only be sent once, we'll need to record this state in this process State.